examples/gno.land/p/demo/timelock/timelock.gno

```diff
@@ -0,0 +1,150 @@
+// Package timelock provides a library for scheduling, cancelling, and
+// executing time-locked operations in Gno. It ensures that
+// operations are only carried out after a specified delay and offers
+// mechanisms for managing and verifying the status of these operations.
+// This package leverages an AVL tree for efficient management of timestamps
+// and integrates role-based access control for administrative tasks.
+//
+// Example Usage:
+//
+//  import "gno.land/p/demo/timelock"
+//  import "gno.land/p/demo/accesscontrol"
+//
+//  Initialize timelock utility with an AVL tree and access control.
+//  timestamps := avl.NewTree()
+//  adminRole := accesscontrol.NewRoleData("admin", std.Address("admin-address"))
+//  timelockUtil := timelock.NewTimelockUtil(timestamps, adminRole)
+//
+//  Schedule an operation with a delay of 60 seconds and a minimum delay of 30 seconds.
+//  id := seqid.ID()
+//  timelockUtil.Schedule(id, 60, 30)
+//
+//  Check if an operation is pending.
+//  isPending := timelockUtil.IsOperationPending(id)
+//
+//  Execute the operation when it is ready.
+//  if timelockUtil.IsOperationReady(id) {
+//      timelockUtil.Execute(id)
+//  }
+//
+//  Update the minimum delay for future operations.
+//  var minDelay int64 = 30
+//  timelockUtil.UpdateDelay(&minDelay, 45)
+
+package timelock // import "gno.land/p/demo/timelock"
+
+import (
+    "strconv"
+    "std"
+    "gno.land/p/demo/avl"
+    "gno.land/p/demo/accesscontrol"
+    "gno.land/p/demo/seqid"
+    "time"
+)
+
+// Marks operations completed
+const DONE_TIMESTAMP = 1
+
+// Represents the status of a planned operation
+type OperationState int
+
+const (
+    Unset OperationState = iota
+    Waiting
+    Ready
+    Done
+)
+
+// TimelockUtil stores the necessary parameters for the timelock operations
+type TimelockUtil struct {
+    timestamps    *avl.Tree
+    accessControl *accesscontrol.RoleData
+}
+
+// New instance of TimelockUtil
+func NewTimelockUtil(timestamps *avl.Tree, accessControl *accesscontrol.RoleData) *TimelockUtil {
+    return &TimelockUtil{
+        timestamps:    timestamps,
+        accessControl: accessControl,
+    }
+}
+
+// Schedules an operation to be carried out after a minimum delay
+func (tl *TimelockUtil) Schedule(id seqid.ID, delay int64, minDelay int64) {
+    if delay < minDelay {
+        panic("timelockutil: Schedule: insufficient delay")
+    }
+    if tl.timestamps.Has(id.Binary()) {
+        panic("timelockutil: Schedule: operation already scheduled")
+    }
+    tl.timestamps.Set(id.Binary(), time.Now().Unix()+delay)
+    std.Emit("CallScheduled", "id", id.String(), "delay", strconv.FormatInt(delay, 10))
+}
+
+// Cancels a planned operation
+func (tl *TimelockUtil) Cancel(id seqid.ID) {
+    if !tl.IsOperationPending(id) {
+        panic("timelock: Cancel: operation not pending")
+    }
+    tl.timestamps.Remove(id.Binary())
```

Comment:
I think there should also be a way to remove completed operations, either by removing the check above and renaming the method, or adding a new method.  If they can never be removed, then it is data stored in a realm that can never be reclaimed short of the realm removing all references to this instance of `TimelockUtil` entirely.

Reply:
Sure, it's good to have a method that allows you to delete an operation, so I've added it.

- https://github.com/gnolang/gno/pull/2307/commits/7f6e8618e8bbdc53fa2e5e162fec7d19b5cc85f6